### DISCUSSION TITLE
FQL - how to solve this ????????

### DISCUSSION THREAD

**Discussion by NSH531** [CrowdStrike/falconpy#713] [Q&A] [answered]
I runned this code but I was getting 400 error   :
```
def hst(myid:str):
    F1=falconpy
    F= falconpy.APIHarness(client_id=client_id,client_secret=secret)
    id_list = 'ID1,ID2,ID3'  # Can also pass a list here: ['ID1', 'ID2', 'ID3']
    response = F1.Hosts(client_id=client_id,client_secret=secret)
    d=(datetime.datetime.now().utcnow()).isoformat()
    F.command()
    lasts=response.query_devices_by_filter(filter=f"last_seen:<={d}*")
  #  return response.query_devices_by_filter(filter=f"hostname:{myid}*")
#.QueryDevicesByFilter(filter=f"host:{myid}*")
                         # )
    return lasts
```

**Comment by jshcodes** [comment, marked as answer]
Hi @NSH531 -

FQL documentation can be found here: https://www.falconpy.io/Usage/Falcon-Query-Language.html

For a [FQL statement using a date filter](https://www.falconpy.io/Usage/Falcon-Query-Language.html#date-syntax), you will need to enclose your date in single quotes.  The wildcard should not be provided.

```python
last_seen_result = hosts.query_devices_by_filter(filter=f"last_seen:<='{d}'")
```

A [FQL statement performing a string search](https://www.falconpy.io/Usage/Falcon-Query-Language.html#string-syntax) would also need to be enclosed in quotes.

```python
# This is a stemmed search
hostname_result = hosts.query_devices_by_filter(filter=f"hostname:'{myid}*'")

# This will find the substring (syntax looks a little strange)
hostname_result = hosts.query_devices_by_filter(filter=f"hostname:*'*{myid}*'"
```

Thank you for the question!

**Reply by NSH531** [reply]
I recieved 500 error:
`
{"body":{"errors":[{"message":"Keyword arguments must be used for this method."}],"resources":[]},"headers":{},"status_code":500}`

**Reply by jshcodes** [reply]
Based upon the question you posted [here](https://github.com/CrowdStrike/falconpy/discussions/714), that 500 is coming from the `__init__` method for the Hosts Service Class, which [only accepts keyword arguments](https://github.com/CrowdStrike/falconpy/blob/main/src/falconpy/_service_class.py#L83). 

The statement `F1.hosts.Hosts.query_devices_by_filter(...)` does not reference the `Hosts` class constructor, nor does it provide keywords to initialize the object. Python is telling you that you are missing input values necessary for creating an instance of the `Hosts` Service Class which must be provided as keywords (not arguments).
